### PR TITLE
[NTOS:KE/x64] Detect CPU vendor properly and store value in PRCB

### PR DESCRIPTION
**Pull request by binarymaster**
Also generate processor identifier properly based on this value
on the Configuration Manager machine-dependent initialization.

Update processor driver INF file accordingly.

JIRA issue: [CORE-17970](https://jira.reactos.org/browse/CORE-17970)

### CODE REVIEW

File: ntoskrnl/config/i386/cmhardwr.c [hunk 1]
```diff
@@ -376,6 +386,32 @@ CmpInitializeMachineDependentConfiguration(IN PLOADER_PARAMETER_BLOCK LoaderBloc
                         (Prcb->CpuStep >> 8),
                         Prcb->CpuStep & 0xff);
             }
+#elif defined(_M_AMD64)
```
Comment (HBelusca) [inline ntoskrnl/config/i386/cmhardwr.c:389]: I was looking at Geoff Chappell's website to see the differences between these `CpuID`, `CpuVendor` etc. fields.
Turns out that `CpuID` field in x64 should always be TRUE: I'm wondering whether we could add somewhere some
`ASSERT(Prcb->CpuID == TRUE);` 🤔

File: ntoskrnl/config/i386/cmhardwr.c [hunk 2]
```diff
@@ -376,6 +386,32 @@ CmpInitializeMachineDependentConfiguration(IN PLOADER_PARAMETER_BLOCK LoaderBloc
...
+            }
+
+            /* Build ID2-style string for all 64-bit CPUs */
+            sprintf(Buffer,
+                    CmpID,
+                    Prcb->CpuType,
+                    (Prcb->CpuStep >> 8),
+                    Prcb->CpuStep & 0xff);
+#else
+            #error Unknown architecture
```
Comment (HBelusca) [inline ntoskrnl/config/i386/cmhardwr.c:413]:
```suggestion
#error Unknown architecture
```

File: ntoskrnl/ke/amd64/cpu.c [hunk 3]
```diff
@@ -89,25 +86,25 @@ KiGetCpuVendor(VOID)
     /* Now check the CPU Type */
     if (!strcmp((PCHAR)Prcb->VendorString, CmpIntelID))
     {
-        return CPU_INTEL;
+        Prcb->CpuVendor = CPU_INTEL;
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/cpu.c:89]: Interesting: the `CpuVendor` field already existed for x64 in NT 5.2, but it had to wait until NT 6.0 to appear in the x86 KPRCB: this should explain why our x86 `KiGetCpuVendor()` doesn't assign anything to this field.

**Review by HBelusca — APPROVED** [review]

File: ntoskrnl/ke/amd64/cpu.c (hunk 3 above)
Comment (binarymaster) [reply ntoskrnl/ke/amd64/cpu.c:89]: Yes, I noticed that too.

File: ntoskrnl/config/i386/cmhardwr.c (hunk 1 above)
Comment (binarymaster) [reply ntoskrnl/config/i386/cmhardwr.c:389]: Don't have a checked build x64 kernel at hand, so don't know if assertion like that exists. Also in [this function](https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/ke/amd64/cpu.c;hb=1f6cd380509e7eab8cb653a748ff947fbf8d6974#l43) the code permanently sets it to TRUE.

Comment (HBelusca) [reply ntoskrnl/config/i386/cmhardwr.c:389]: ah ok then no need for any assertion.

File: ntoskrnl/config/i386/cmhardwr.c [hunk 4]
```diff
@@ -357,6 +366,7 @@ CmpInitializeMachineDependentConfiguration(IN PLOADER_PARAMETER_BLOCK LoaderBloc
             ConfigData.ComponentEntry.AffinityMask = AFFINITY_MASK(i);
             ConfigData.ComponentEntry.Identifier = Buffer;
 
+#if defined(_M_IX86)
             /* Check if the CPU doesn't support CPUID */
             if (!Prcb->CpuID)
```
Comment (Extravert-ir) [inline ntoskrnl/config/i386/cmhardwr.c:371]: We don't support CPUs without CPUID, so this should be removed completely

**Comment by Extravert-ir** [comment]
I don't see the point - why complicate the code and detect CPU vendor for that string? Where is it going to be used?
They share the same instruction set

File: ntoskrnl/config/i386/cmhardwr.c [hunk 5]
```diff
@@ -241,6 +247,9 @@ CmpInitializeMachineDependentConfiguration(IN PLOADER_PARAMETER_BLOCK LoaderBloc
     CPU_INFO CpuInfo;
     ULONG ExtendedId;
     PKPRCB Prcb;
+#ifdef _M_AMD64
+    PCHAR CmpID;
```
Comment (Extravert-ir) [inline ntoskrnl/config/i386/cmhardwr.c:251]: I don't see the point - why complicate the code and detect CPU vendor for that string? Where is it going to be used?
They share the same instruction set

**Review by Extravert-ir — CHANGES_REQUESTED** [review]

File: ntoskrnl/config/i386/cmhardwr.c (hunk 5 above)
Comment (HBelusca) [reply ntoskrnl/config/i386/cmhardwr.c:251]: what string?

File: ntoskrnl/config/i386/cmhardwr.c (hunk 4 above)
Comment (binarymaster) [reply ntoskrnl/config/i386/cmhardwr.c:371]: I didn't touched that code anyway :wink:

File: ntoskrnl/config/i386/cmhardwr.c (hunk 5 above)
Comment (binarymaster) [reply ntoskrnl/config/i386/cmhardwr.c:251]: This string affects ACPI identifiers, and they affect what CPU driver will be assigned to the device. We don't have specialized CPU drivers though, but this is to match Windows behavior.

Comment (Extravert-ir) [reply ntoskrnl/config/i386/cmhardwr.c]: So we are going to have one driver for all CPUs on 32bit and 3 drivers on 64bit? Where is the logic here?

Comment (binarymaster) [reply ntoskrnl/config/i386/cmhardwr.c]: We will have one driver for both cases, until that changes for some reason. Note that Windows also have different CPU drivers on 32-bit.

**Review by HBelusca — APPROVED** [review]

**Pull request merged**